designs/auto-create-ecr.md

```diff
@@ -0,0 +1,306 @@
+    stack_name[:104] + "-" + stack_md5[:8] + "-" + "CompanionStack"
+    #104 + 1 + 8 + 1 + 14 = 128 max char
+    
+repo_output_logical_id = 
+    lambda_logical_id[:52] + function_md5 [:8] + "Out"
+    #52 + 8 + 3 = 63 max char
+
+Exmaple:
+    Input:
+        Customer Stack Name: Hello-World-Stack
+        Function 1 Logical ID: TestFunction01
+        Function 2 Logical ID: AnotherTestFunction02
+    Output:
+        Companion Stack Name: Hello-World-Stack-925976eb-CompanionStack
+        Function 1 Repo Logical ID: TestFunction0150919004Repo
+        Function 1 Repo Name: helloworldstack925976eb/testfunction0150919004repo
+        Function 2 Repo Logical ID: AnotherTestFunction025c2cfd8cRepo
+        Function 2 Repo Name: helloworldstack925976eb/anothertestfunction025c2cfd8crepo
+```
+
+**Companion Stack Structure**
+```
+AWSTemplateFormatVersion : '2010-09-09'
+Transform: AWS::Serverless-2016-10-31
+Description: AWS SAM CLI Managed ECR Repo Stack
+Metadata:
+  SamCliInfo: 1.18.0
+  CompanionStackname:  Hello-World-Stack-925976eb-CompanionStack
+
+Resources:
+
+  TestFunction0150919004Repo:
+    Type: AWS::ECR::Repository
+    Properties:
+      RepositoryName: helloworldstack925976eb/testfunction0150919004repo
+      Tags:
+        - Key: ManagedStackSource
+          Value: AwsSamCli
+        - Key: AwsSamCliCompanionStack
+          Value: Hello-World-Stack-925976eb-CompanionStack
+
+      RepositoryPolicyText: 
+        Version: "2012-10-17"
+        Statement:
+          -
+            Sid: AllowLambdaSLR
+            Effect: Allow
+            Principal:
+              Service:
+                - "lambda.amazonaws.com"
+            Action:
+                - "ecr:GetDownloadUrlForLayer"
+                - "ecr:GetRepositoryPolicy"
+                - "ecr:BatchGetImage"
+
+  AnotherTestFunction025c2cfd8cRepo:
+    Type: AWS::ECR::Repository
+    Properties:
+      RepositoryName: helloworldstack925976eb/anothertestfunction025c2cfd8crepo
+      Tags:
+        - Key: ManagedStackSource
+          Value: AwsSamCli
+        - Key: AwsSamCliCompanionStack
+          Value: Hello-World-Stack-925976eb-CompanionStack
+
+      RepositoryPolicyText: 
+        Version: "2012-10-17"
+        Statement:
+          -
+            Sid: AllowLambdaSLR
+            Effect: Allow
+            Principal:
+              Service:
+                - "lambda.amazonaws.com"
+            Action:
+                - "ecr:GetDownloadUrlForLayer"
+                - "ecr:GetRepositoryPolicy"
+                - "ecr:BatchGetImage"
+
+Outputs:
+
+  None:
+    Value: !Sub ${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/${TestFunction0150919004Repo}
+
+  None:
+    Value: !Sub ${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/${AnotherTestFunction025c2cfd8cRepo}
+```
+
+Documentation Changes
+=====================
+* New option `--resolve-image-repos`. This option will auto create repos without the needs of going through guided experience. 
```

Comment:
should we showcase how non interactive usage looks like when specifying `--resolve-image-repos`. I'm assuming it does no destructive actions as opposed to sam deploy --guided which prompts to delete repos when functions dont exist.

Reply:
--resolve-image-repos will be the same as guided experience including destructive actions.

1. This will keep a consistent behavior as the guided. In guided, SAM CLI will abort deployment if deletion of auto created image repos is denied.
2. For UX, this will avoid image repos and functions mapping to an invalid state where orphaned image repos exist. For this case, we will also need to track which repos should be kept and makes the sync less robust.
3. From a security perspective, keeping old image repos will increase the impact radius of an information leakage. A customer might expect a sensitive image repo will be deleted as soon as the function itself is also removed like the guided experience.